File: base/src/main/java/com/windscribe/vpn/api/response/PortMapResponse.java
/*
 * Copyright (c) 2021 Windscribe Limited.
 */

package com.windscribe.vpn.api.response;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import java.util.List;

@SuppressWarnings("unused")
public class PortMapResponse {

    @SerializedName("portmap")
    @Expose
    private final List<PortMap> portmap = null;
    @SerializedName("suggested")
    @Expose
    private Suggested suggested;

    public List<PortMap> getPortmap() {
        return portmap;
    }

    @Nullable
    public Suggested getSuggested() {
        return suggested;
    }

    public boolean isProtocolSuggested() {
        return suggested != null;
    }

    @NonNull
    @Override
    public String toString() {
        return "PortMapResponse{" +
                ", portmap=" + portmap +
                ", suggested=" + suggested +
                '}';
    }

    public static class PortMap {

        @SerializedName("legacy_ports")
        @Expose
        private final List<String> legacyPorts = null;
        @SerializedName("ports")
        @Expose
        private final List<String> ports = null;
        @SerializedName("heading")
        @Expose
        private String heading;
        @SerializedName("protocol")
        @Expose
        private String protocol;

        @SerializedName("use")
        @Expose
        private String use;

        public String getHeading() {
            return heading;
        }

        public List<String> getLegacyPorts() {
            return legacyPorts;
        }

        public List<String> getPorts() {
            return ports;
        }

        public String getProtocol() {
            return protocol;
        }

        public String getUse() {
            return use;
        }

        @NonNull
        @Override
        public String toString() {
            return "PortMap{" +
                    "protocol='" + protocol + '\'' +
                    ", heading='" + heading + '\'' +
                    ", use='" + use + '\'' +
                    ", ports=" + ports +
                    ", legacyPorts=" + legacyPorts +
                    '}';
        }
    }

    public static class Suggested {

        @SerializedName("port")
        @Expose
        private int port;

        @SerializedName("protocol")
        @Expose
        private String protocol;

        public int getPort() {
            return port;
        }

        public String getProtocol() {
            return protocol;
        }

        @NonNull
        @Override
        public String toString() {
            return "Suggested{" +
                    "protocol='" + protocol + '\'' +
                    ", port='" + port + '\'' +
                    '}';
        }
    }

}

File: base/src/main/java/com/windscribe/vpn/api/response/PortMapResponse.kt
/*
 * Copyright (c) 2021 Windscribe Limited.
 */

package com.windscribe.vpn.api.response

import androidx.annotation.Keep
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName

@Keep
class PortMapResponse {
    @SerializedName("portmap")
    @Expose
    val portmap: List<PortMap>? = null

    @SerializedName("suggested")
    @Expose
    val suggested: Suggested? = null

    val isProtocolSuggested: Boolean
        get() = suggested != null

    override fun toString(): String =
        "PortMapResponse{" +
            ", portmap=" + portmap +
            ", suggested=" + suggested +
            '}'

    @Keep
    class PortMap {
        @SerializedName("legacy_ports")
        @Expose
        val legacyPorts: List<String>? = null

        @SerializedName("ports")
        @Expose
        val ports: List<String>? = null

        @SerializedName("heading")
        @Expose
        val heading: String? = null

        @SerializedName("protocol")
        @Expose
        val protocol: String? = null

        @SerializedName("use")
        @Expose
        val use: String? = null

        override fun toString(): String =
            "PortMap{" +
                "protocol='" + protocol + '\'' +
                ", heading='" + heading + '\'' +
                ", use='" + use + '\'' +
                ", ports=" + ports +
                ", legacyPorts=" + legacyPorts +
                '}'
    }

    @Keep
    class Suggested {
        @SerializedName("port")
        @Expose
        val port: Int = 0

        @SerializedName("protocol")
        @Expose
        val protocol: String? = null

        override fun toString(): String =
            "Suggested{" +
                "protocol='" + protocol + '\'' +
                ", port='" + port + '\'' +
                '}'
    }
}
